배포 환경 준비

wanniDev opened the pull request:
이번 pr은 클라우드 환경에 서버를 배포를 위한 사전 준비 및 간단한 리팩토링을 수행하였습니다. 이번 작업의 주요 포인트는 다음과 같습니다.

## 1. 카프카 템플릿 wrapperClass 구성: 
 
테스트용 인터페이스를 구분하기위해 `StringKafkaTemplateWrapper`를 활용했습니다. 멘토링 세션에서 논의해본대로 `KafkaTemplate`을 직접 mocking하는 방향도 고려해보았으나, 이 방법의 경우 빌드시 진행되는 카프카 클라이언트의 설정 체크 및 헬스체크 등에 대한 mock 작업을 수행해야했습니다.
카프카 클러스터 및 브로커 서버에 대한 테스트 시나리오가 없는 시점에서는 단순한 방향을 선택하는 것이 좋을 것 같습니다.

## 2. TossPaymentsTransactionEventPublisher

결제 승인 요청 및 주문 결제 발행시 활용되는 로깅 이벤트를 처리하는 부분을 분리하였습니다. SOLID 원칙의 단일 책임 원칙을 의식하였는데요.
정말 단순히 '클래스의 역할'을 하나로 분류하는 것은 주관적인 요소가 많이 개입할 수 있기에 제 경우에는 클래스 하나의 코드 길이기 100줄을 넘어가는 시점에서부터 SRP를 고려하고 있습니다.

## 3. 배포 환경 구서

아직도 러프한 상태지만, 적어도 소스코드를 클라우드 인스턴스에 띄우고 도메인 주소를 입력하면 바로 접속할 수 있는 환경을 구성하는 것에 집중했습니다.

도커 설정, 도메인 구매, dns 적용 및 ssl 인증서 등록 등의 작업을 진행하였으며, ssl 암호와 같은 민감한 정보들은 프로퍼티 파일을 활용하되 깃헙과 같은 퍼블릭한 환경에서는 해당 파일을 아예 공개하지 않는 방향으로 진행하였습니다. 추가로 전반적으로 api 시그니처에 대한 수정 작업도 진행했습니다.(서브도메인에 api가 포함되었으므로, 전체 시그니처 url에 'api' 문구 삭제)

wanniDev committed:
Chore: 사용하지 않는 app 컴포즈 구성 삭제

wanniDev committed:
Rename: mysql 구성하는 컴포즈의 기능에 맞게 파일이름 변경

wanniDev committed:
Remove: 하드코딩되어 있는 의미없는 api 삭제

wanniDev committed:
Test: 주문 결제 발행 api 변경된 상태코드 반영

wanniDev committed:
Remove: 이제는 사용 안하는 화면 렌더링용 컨트롤러 삭제

도메인을 구매하여 DNS 서버에 띄웠는데, 이제 api 서버는 api.domain.com 로 구성하기로 했다.
따라서 더이상 화면 렌더링용 스프링의 컨트롤러는 불필요하다고 판단하여 삭제하기로 했다.

wanniDev committed:
Refactor: api.domain.com 패턴으로  api 시그니처 수정

이미 도메인에 api 라는 부분이 명시되어있으므로 더이상 요청 url에 api를 포함할 필요가 없다.

wanniDev committed:
Refactor: 가독성 향상 및 원할한 테스트를 위해 카프카 이벤트 생성 및 로그 기록 코드 리팩토링

wanniDev committed:
Feat: https로 배포하는데 필요한 ssl 키스토어 파일 추가

도메인은 가비아
dns 등록은 ncp
ssl은 Let's encrypt를 활용하였다.

wanniDev committed:
Refactor: 불필요한 세미클론 정리

wanniDev committed:
Feat: 원활한 kafkaTemplate 테스트를 위해 wrapper 인터페이스와 기본 구현체 구성

Wrapper 클래스를 활용하여 개발, 프로덕션 환경에는 실제 카프카 브로커 서버를 사용하고 테스트를 할때는 별도의 모킹을 활용하고자 한다.

wanniDev committed:
Test: 테스트용 mock 카프카템플릿 구성

wanniDev committed:
Refactor: 다양한 형태의 프론트 구성에 대응할 수 있도록 회원가입 확인 api에 사용되는 리다이렉트 url을 프로퍼티 변수로 변경

wanniDev committed:
Refactor: 외부 api를 최대한 통제된 상태에서 사용할 수 있도록 결제요청은 퍼블릭한 환경에서는 관리자만 허용

wanniDev committed:
Feat: pg 이용이 가능한 관리자 등록 api 구현

wanniDev committed:
Fix: 시연환경에서 실제로 사용할 수 있는 빈으로 의존성 변경

wanniDev committed:
Test: 관리자 회원등록 api 테스트코드 반영

wanniDev committed:
Document: 관리자 회원가입 api rest docs 작성

wanniDev committed:
Hotfix: 관리자 등록시 admin role 적용 및 인가 애노테이션 추가

wanniDev committed:
Test: 인가 체크가 적용된 테스트 반영

f-lab-troy commented on account-api/account-domain/src/main/kotlin/org/collaborators/paymentslab/account/domain/Account.kt:
```diff
@@ -58,7 +59,8 @@ class Account protected constructor(
 
     fun isValidToken(token: String): Boolean {
-        return this.emailCheckToken.equals(token)
+//        return this.emailCheckToken.equals(token)
```
추후에 수정하는 것으로 이해하겠습니다

wanniDev replied:
로컬에서만 프로젝트를 구성했다면, 그냥 망설임 없이 이메일 인증 서비스를 구현했을텐데, 직접 도메인을 내고 공개된 환경에서 적용하려니 괜히 생각이 많아지네요... 이메일 인증 서비스 부분은 리스크 파악을 충분히 진행한 뒤에 적용해보도록 하겠습니다. :)

f-lab-troy commented on account-api/account-domain/src/main/kotlin/org/collaborators/paymentslab/account/domain/AccountRegister.kt:
```diff
@@ -1,17 +1,19 @@
             throw DuplicatedEmailException()
 
-        val account = Account.register(email, encrypt.encode(password), username, phoneNumber)
+        val account = Account.register(email, encrypt.encode(password), username, phoneNumber, roles)
```
password가 encode되기 전에 로깅을 한다면 어떤 위험성이 있고 막을 수 있는 방법이 있는지도 생각해보시면 좋습니다

wanniDev replied:
우선 password가 유출될 것 같습니다. 해커가 해당 정보를 탈취했을 때, 보안의 취약점이 생기는 경우도 있지만.. 이 서비스를 운영하는 내부 인원도 로그에 접근할 수 있겠습니다.

많은 고민을 해보았습니다만, 우선은 암호화가 안된 비밀번호를 '실수로' 출력하는 경우를 대비하고자 해당 dto와 entity의 toString()을 재정의하여 비밀번호가 그대로 출력되는 경우를 막았습니다.

그리고, debug 로그 출력으로 password를 출력하게 하고 debug 로그 출력 레벨을 유지한 상태로 prod 환경에 배포하는 경우를 배제하기 위해 애플리케이션을 실행하기 전에 프로파일과 로그 출력 레벨을 검사하여, 경우에 따라 애플리케이션을 종료시키는 방법을 채택하였습니다.

그럼에도 불구하고.. 빌런(?) 개발자가 일부러 비밀번호를 로그로 출력하게하는 경우도 있을것 같은데.. 정적 코드 분석 도구 등을 활용하여 해당 출력 코드가 있으면 배포를 못하게 하는 방법이 있을 것 같은데, 이 부분은 좀 더 고민해봐야할 것 같습니다.

f-lab-troy replied:
서비스에 들어오는 가장 앞단에서부터 암호를 보호하는 것이 가장 좋은 방법일거라 생각합니다

f-lab-troy commented on account-api/account-presentation/src/main/kotlin/org/collaborators/paymentslab/account/presentation/AuthenticationApi.kt:
```diff
@@ -1,26 +1,42 @@
+    @PostMapping("register/admin")
+    fun registerAdmin(@RequestBody @Valid request: RegisterAdminAccountRequest) {
+        accountService
+            .registerAdmin(
```
registerAdmin안에 RegisterAdminAccount을 넘겨주는 것이 가독성면으로 볼때 부자연스럽다고 생각 할 수도 있을 것 같습니다.

wanniDev replied:
네, 저도 동의합니다.. 부여할 권한만 다를 뿐인데, dto를 분리하는건 가독성 뿐만 아니라, 회원가입과 관련된 코드를 수정하는데 드는 유지비용도 증가할 것 같습니다.

똑같은 dto를 쓰되, 관리자 회원가입만 생성자 파라미터에 admin role을 추가하는 것으로 진행해보았습니다.

f-lab-troy commented on app/src/test/kotlin/org/collaborators/paymentslab/account/presentation/AuthenticationApiTest.kt:
```diff
@@ -32,7 +33,7 @@ class AuthenticationApiTest @Autowired constructor(
         this.mockMvc.perform(
             RestDocumentationRequestBuilders
-                .post("$V1_API_AUTH/register")
+                .post("$V1_AUTH/register")
```
테스트를 위해 api path들을 상수의 집합으로 만들어 놓아도 중복을 막을 수 있을 것 같습니다.

wanniDev replied:
혹시 'register, reissurance'와 같이 반복되는 부분들도 상수화하여 중복을 줄였으면 좋겠다로 이해해도 될까요?

f-lab-troy commented on payment-api/payment-application/src/main/kotlin/org/collaborators/paymentslab/payment/application/PaymentOrderService.kt:
```diff
@@ -10,7 +10,7 @@ import org.springframework.transaction.annotation.Transactional
-    fun generate(command: PaymentOrderCommand): String {
-        return paymentOrderProcessor.process(command.accountId, command.orderName, command.amount)
+    fun generate(command: PaymentOrderCommand) {
+        paymentOrderProcessor.process(command.accountId, command.orderName, command.amount)
```
command자체를 넘겨주면 생길 수 있는 문제가 있을까요

wanniDev replied:
현재 프로젝트에서 application 계층은 domain 계층을 의존하되, domain 계층은 프레임워크를 의존하지 않는 독립된 계층을 유지하기 위해, common을 제외한 다른 계층은 전혀 의존하지 않고 있습니다.

따라서, 지금 기준으로 command를 넘겨주면 domain 계층에 속하는 paymentOrderProcessor는 command 자체를 인식하지 못해 컴파일 에러가 발생합니다.

물론, 프로젝트 구조 설정을 바꿔도 기능상으로는 문제가 없습니다. 다만, 모듈러 모놀리식 기반의 layered architecture에서 계층별 의존관계에 대한 일관성이 깨질 수 있습니다.

그러나 command가 넘겨주는 필드의 수가 커질수록 코드의 길이도 불필요하게 늘어나는 현상이 발생하는 문제가 있기에.. 초기에 생각했던 layered architecture의 의존방향을 수정해야하는지 계속 고민중에 있습니다... 혹시 이 부분에 대해 힌트라도 구할 수 있다면 정말 감사하겠습니다. :)

f-lab-troy replied:
> paymentOrderProcessor는 command 자체를 인식하지 못해 컴파일 에러가 발생합니다.

이름을 보아도 PaymentOrderCommand와 paymentOrderProcessor는 밀접한 연관이 있어보입니다.
공유가 가능하다고 생각되는 Command를 common에 포함시켜주면 생기는 문제가 있을지에 대한 고민을 하시고 접근하시면 좋을 것 같습니다.

f-lab-troy commented on payment-api/payment-infrastructure/src/main/kotlin/org/collaborators/paymentslab/payment/infrastructure/tosspayments/TossPaymentsTransactionEventPublisher.kt:
```diff
@@ -0,0 +1,54 @@
+        val principal = SecurityContextHolder.getContext().authentication.principal as AuthenticatedUser
+        val newPaymentOrder = PaymentOrder.newInstance(accountId, orderName, amount)
+        val paymentOrder = paymentOrderRepository.save(newPaymentOrder)
+        if (accountId != principal.id) {
```
principal check의 반복을 막을 수 있는 방법을 고려해보시면 좋을 것 같습니다

wanniDev replied:
`@AuthenticationPrincipal`을 활용해보는 방법도 생각해보았는데, 그러면 api 컨트롤러 코드에 반복이 일어나고, 기존 함수들을 전부 수정해야하는 상황이 발생하여..

principal을 호출하는 부분에 대한 반복은 확장함수를 활용하요 줄여보았습니다.

다만, principal을 호출하고 비교하는 부분에 대해서는 아직은 일정하게 반복되는 패턴을 찾지는 못한 것 같습니다. 해당 부분은 틈틈이 모니터링 하면서 리팩토링을 진행해보도록 하겠습니다.

wanniDev committed:
Refactor: 코틀린의 특성을 활용하여 기존의 java 스타일로 장황하게 작성된 엔티티 모델 코드를 좀 더 간결하게 리팩토링 수행

data class를 활용하면 기존의 equals, hashCode를 활용하지 않아도 되지만, jpa 엔티티 클래스의 경우 다른 엔티티와의 연관관계에 따라 메모리 누수 및
순환 참조로 인한 스택 오버플로우 현상이 일어날 수 있다.

wanniDev committed:
Remove: 사용하지 않는 컨트롤러 코드 삭제

wanniDev committed:
Feat: 회원가입 시 활용되는 비밀번호가 인코딩되기 이전에 출력되는 경우를 막기 위해, toString(), 커스텀 setter 변경

인스턴스 자체를 출력하면서, 비밀번호가 유출되는 경우를 막기 위해 toString()을 재정의 하면서 password를 '[PROTECTED]'로 하드코딩 처리
정해진 Account 내부 함수외로 password가 변경되는 경우를 방지하기 위해 setter의 접근제어자를 private로 변경

wanniDev committed:
Feat: prod 환경에서는 로그 출력레벨을 debug로 지정하지 못하게 하는 코드 추가

wanniDev committed:
Refactor: 가독성을 위해 권한만 다를 뿐 프로세스는 동일한 회원가입의 dto를 하나로 통일

추가로 이메일 인증이 완료되면 accountRegister.registerConfirm 코드는 삭제해야한다.

wanniDev committed:
Refactor: 반복적으로 사용되는 principal 호출을 코틀린의 확장함수를 통해 재사용하는 방향으로 수정

wanniDev committed:
Feat: 정해진 프로파일만 사용하여 프로파일 지정에 대한 일관성 유지

wanniDev committed:
Remove: 사용하지 않는 휴대폰 인증 api 삭제

wanniDev committed:
Refactor: 하드코딩의 반복과 타이핑 실수를 줄이기 위해 api 시그니처를 상수로 변경

앞으로 api 시그니처가 변경될 때마다 ApiEndpoints의 내용을 바꾸도록 한다.

f-lab-troy commented on account-api/account-domain/src/main/kotlin/org/collaborators/paymentslab/account/domain/Account.kt:
```diff
@@ -57,9 +57,9 @@ class Account protected constructor(
-        return this.emailCheckToken.equals(token)
-    }
+    fun isValidToken(token: String) = true
+//    = token == emailCheckToken
```
추후 체크하는 것으로 알겠습니다

f-lab-troy commented on account-api/account-domain/src/main/kotlin/org/collaborators/paymentslab/account/domain/Account.kt:
```diff
@@ -75,4 +75,22 @@ class Account protected constructor(
     override fun hashCode(): Int {
         return id?.hashCode() ?: 0
```
hashCode가 0일 경우 어떤 문제가 생길 수 있는지에 대한 생각을 해보시면 좋습니다

f-lab-troy commented on account-api/account-presentation/src/main/kotlin/org/collaborators/paymentslab/account/presentation/request/RegisterAdminAccountRequest.kt:
```diff
@@ -0,0 +1,19 @@
+    val password: String,
+    @field:NotBlank
+    val username: String,
+    @field:NotBlank @field: Pattern(regexp = "^01(?:0|1|[6-9])-(?:\\d{3}|\\d{4})-\\d{4}$")
```
regex는 가독성이 떨어지기 때문에 예제나 풀이를 코멘트로 달아주시면 좋습니다

f-lab-troy commented on app/src/main/kotlin/org/collaborators/paymentslab/PaymentsLabApplication.kt:
```diff
@@ -1,11 +1,48 @@
+				|| applicationLoggingLevel.equals("DEBUG", ignoreCase = true)
+
+	private fun isProfileInBoundary(env: ConfigurableEnvironment)
+		= isProfileLocal(env) || isProfileTest(env) || isProfileDev(env) || isProfileProduction(env)
```
간단한 로직이라 상관없지만 어떤 코드든 성능을 고려해서 생각해보면 좋습니다

f-lab-troy commented on app/src/main/kotlin/org/collaborators/paymentslab/PaymentsLabApplication.kt:
```diff
@@ -1,11 +1,48 @@
+	private fun isProfileInBoundary(env: ConfigurableEnvironment)
+		= isProfileLocal(env) || isProfileTest(env) || isProfileDev(env) || isProfileProduction(env)
+
+	private fun isProfileProduction(env: ConfigurableEnvironment) = "prod" in env.activeProfiles
```
magic string은 최대한 줄이고 상수나 enum을 사용하는 것이 좋습니다

f-lab-troy commented on app/src/test/kotlin/org/collaborators/paymentslab/account/presentation/AuthenticationApiTest.kt:
```diff
@@ -62,6 +63,48 @@ class AuthenticationApiTest @Autowired constructor(
+
     @Test
     @DisplayName("잘못된 회원가입 폼 입력으로 인한 오류 발생 테스트")
     fun registerErrorTest() {
```
아래 테스트 메서드들과 공통된 코드를 추출하여 중복을 막을 수 있을지에 대한 고민을 해보시면 좋을 것 같습니다

f-lab-troy commented on common/src/main/kotlin/org/collaborator/paymentlab/common/Constants.kt:
```diff
@@ -6,9 +6,6 @@ const val URI_HOST = "127.0.0.1"
-// FRONT PAGE SIGNATURE(TEMP)
-const val V1_TOSS_PAYMENTS = "/v1/toss-payments"
\ No newline at end of file
+//const val V1_ACCOUNT = "/v1/account"
```
필요없는 코드는 git history에 남아있기때문에 항상 지워주는 습관이 좋습니다.

f-lab-troy commented on payment-api/payment-infrastructure/src/main/kotlin/org/collaborators/paymentslab/payment/infrastructure/tosspayments/TossPaymentsPaymentOrderProcessor.kt:
```diff
@@ -1,44 +1,12 @@
-        }
-        return newPaymentOrder.id.toString()
+    override fun process(accountId: Long, orderName: String, amount: Int) {
+        paymentsTransactionEventPublisher.publishAndRecord(accountId, orderName, amount)
```
이 Processor가 하는 것의 전부가 Publisher를 같은 변수로 부르는 것이라면 caller가 publisher를 직접 부르는 것에 대한 고려도 해보면 좋습니다
